Framework/HistogramData/test/HistogramRebinTest.h

```diff
@@ -0,0 +1,324 @@
+#ifndef MANTID_HISTOGRAMDATA_HISTOGRAMREBINTEST_H_
+#define MANTID_HISTOGRAMDATA_HISTOGRAMREBINTEST_H_
+
+#include <cxxtest/TestSuite.h>
+
+#include "MantidHistogramData/Histogram.h"
+#include "MantidHistogramData/HistogramRebin.h"
+#include "MantidHistogramData/LinearGenerator.h"
+
+#include <algorithm>
+#include <random>
+
+using namespace Mantid::HistogramData;
+
+class HistogramRebinTest : public CxxTest::TestSuite {
+public:
+  // This pair of boilerplate methods prevent the suite being created statically
+  // This means the constructor isn't called when running other tests
+  static HistogramRebinTest *createSuite() { return new HistogramRebinTest(); }
+  static void destroySuite(HistogramRebinTest *suite) { delete suite; }
+
+  void testExecRebinCounts() {
+    TS_ASSERT_THROWS_NOTHING(rebinCounts(
+        getCountsHistogram(), BinEdges(10, LinearGenerator(0, 0.5))));
+  }
+
+  void testExecRebinFrequency() {
+    TS_ASSERT_THROWS_NOTHING(rebinFrequencies(
+        getFrequencyHistogram(), BinEdges(10, LinearGenerator(0, 0.5))));
+  }
+
+  void testRebinFrequenciesFailModes() {
+    BinEdges edges(5, LinearGenerator(0, 2));
+    // Incorrect YMode
+    TS_ASSERT_THROWS(rebinFrequencies(getCountsHistogram(), edges),
+                     std::runtime_error);
+    // Incorrect XMode
+    TS_ASSERT_THROWS(rebinFrequencies(Histogram(Histogram::XMode::Points,
+                                                Histogram::YMode::Frequencies),
+                                      edges),
+                     std::runtime_error);
+    // No YMode set
+    TS_ASSERT_THROWS(
+        rebinFrequencies(Histogram(BinEdges(10, LinearGenerator(0, 0.5))),
+                         edges),
+        std::runtime_error);
+  }
+
+  void testRebinCountsFailModes() {
+    BinEdges edges(5, LinearGenerator(0, 2));
+    // Incorrect YMode
+    TS_ASSERT_THROWS(rebinCounts(getFrequencyHistogram(), edges),
+                     std::runtime_error);
+    // Incorrect XMode
+    TS_ASSERT_THROWS(rebinCounts(Histogram(Histogram::XMode::Points,
+                                           Histogram::YMode::Counts),
+                                 edges),
+                     std::runtime_error);
+    // No YMode set
+    TS_ASSERT_THROWS(
+        rebinCounts(Histogram(BinEdges(10, LinearGenerator(0, 0.5))), edges),
+        std::runtime_error);
+  }
+
+  void testRebinFailsBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 3, 5, 7};
+    BinEdges edges(binEdges);
+
+    TS_ASSERT_THROWS(rebinCounts(getCountsHistogram(), edges),
+                     std::runtime_error);
+    TS_ASSERT_THROWS(rebinFrequencies(getFrequencyHistogram(), edges),
+                     std::runtime_error);
+  }
+
+  void testRebinFailsInputBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 3, 5, 7};
+    Histogram hist(BinEdges(std::move(binEdges)), Counts(5, 10));
+    BinEdges edges{1, 2, 3, 4, 5, 6};
```

Comment:
Since `VectorHelper::rebin()` failed differently depending on whether the bad bin edges coincided in input and output or not, maybe we should test both cases here and in the test above this one.